slo_export.py

```diff
@@ -0,0 +1,232 @@
+    return config
+
+
+def get_slo_configs(config):
+    """Connect to Datadog using our configuration and retrieve all SLO configs.
+
+    Arguments:
+        config (dict): A dictionary containing all required program
+            configuration options.
+
+    Returns:
+        slo_configs (dict): A dictionary containing all retrieved SLO configs.
+    """
+    api_config = Configuration()
+    api_config.api_key['apiKeyAuth'] = config['DD_API_KEY']
+    api_config.api_key['appKeyAuth'] = config['DD_APP_KEY']
+    api_config.server_variables['site'] = config['DD_SITE']
+    with ApiClient(api_config) as api_client:
+        api_instance = ServiceLevelObjectivesApi(api_client)
+        response = api_instance.list_slos()
+    slo_configs = response.to_dict()
+    return slo_configs
+
+
+def normalize_name(name):
+    """Turn a displayname in a kubernetes-stylei/RFC 1123-compliant name."""
+    name = name.lower()
+    name = re.sub('[^a-zA-Z0-9- ]', '', name)
+    name = re.sub('\s+', '-', name)
+    name = re.sub(r'(-)+', r'\1', name)
+    name = name[:63]
+    name = name.strip('-')
+    return name
+
+
+def escape_chars(description):
+    """Datadog description strings might contain quotes or newlines."""
+    description = re.sub('\n', ' ', description)
+    description = re.sub('\"', '\\\"', description)
+    return description
+
+
+def extract_values(slo, config):
+    """Construct a dictionary containing relevant values from a single DD SLO.
+
+    For a better understanding of the complexity of the values being extracted,
+    please see the definition of the SLOResponse class in datadog_api_client.v1.
+
+    Arguments:
+        slos (dict): A dictonary containing an entire DD SLO config and
+            metadata.
+        config (dict): Additional configuration specifying Nobl9-specific
+            values.
+
+    Returns:
+        slos_values (dict): A dictionary with values ready to populate Nobl9
+            YAML templates.
+    """
+    slo_values = {}
+    slo_values['name'] = normalize_name(slo['name'])
+    slo_values['displayName'] = slo['name'][:63]
+    slo_values['description'] = escape_chars(slo['description'])
+    slo_values['datasource'] = config['N9_DS']
+    slo_values['datasource_project'] = config['N9_DS_PROJECT']
+    slo_values['kind'] = config['N9_DS_KIND']
+    slo_values['project'] = config['N9_PROJECT']
+    slo_values['service_name'] = slo_values['name']
+    slo_values['good'] = slo['query']['numerator']
+    slo_values['total'] = slo['query']['denominator']
+    slo_values['thresholds'] = []
+
+    num_thresholds = len(slo['thresholds'])
+    for threshold in range(num_thresholds):
+        target = slo['thresholds'][threshold]['target']
+        slo_values['thresholds'].append({})
+        slo_values['thresholds'][threshold]['budgetTarget'] = target
+        display_name = slo['thresholds'][threshold]['target_display']
+        slo_values['thresholds'][threshold]['displayName'] = display_name
```

Comment:
I find version with list iterator slightly easier to read, wdyt?
```suggestion
    for dd_threshold in slo["thresholds"]:
        n9_threshold = {}
        n9_threshold["budgetTarget"] = dd_threshold["target"]
        n9_threshold["displayName"] = dd_threshold["target_display"]
        slo_values["thresholds"].append(n9_threshold)
```

Reply:
I left this as-is for now. I know the whole function is complicated to read, but I think I'd like to leave cleaning it up to a potential complete overhaul for readability.